rubicon/java/android_events.py

```diff
@@ -168,45 +193,125 @@ def _run_scheduled_callbacks_then_restart_loop(self):
-        # Create a Java Runnables to tick the event loop forward. We statically create
-        # Runnables to avoid creating lots of temporary objects.
-        python_self = self
 
-        class RunScheduledCallbacksThenRestartLoopRunnable(Runnable):
-            def run(self):
-                python_self._run_scheduled_callbacks_then_restart_loop()
+class AndroidInterop:
+    """Encapsulate details of Android event loop cooperation."""
+
+    def __init__(self):
+        self._runnable_by_key = {}
+        self._calling_soon = set()
+        self._handler = None
+
+    @property
+    def handler(self):
+        if self._handler is None:
+            # We use `android.os.Handler.postDelayed()` to ask for wakeup. This requires an instance
+            # of `Handler`, which we cache. We use the default constructor which assumes we are on
+            # the Android UI thread.
+            self._handler = Handler()
+        return self._handler
+
+    def get_or_create_runnable(self, key, fn):
+        if key in self._runnable_by_key:
+            return self._runnable_by_key[key]
+        android_interop_self = self
 
-        class RunEventLoopOnceCooperativelyRunnable(Runnable):
+        class PythonRunnable(Runnable):
             def run(self):
-                python_self._run_event_loop_once_cooperatively()
-
-        self._runnables = {
-            "_run_scheduled_callbacks_then_restart_loop": RunScheduledCallbacksThenRestartLoopRunnable(),
-            "_run_event_loop_once_cooperatively": RunEventLoopOnceCooperativelyRunnable(),
-        }
-        self._run_event_loop_once_cooperatively()
-        print("Started event loop")
+                if key in android_interop_self._calling_soon:
+                    android_interop_self._calling_soon.remove(key)
+                fn()
+
+        self._runnable_by_key[key] = PythonRunnable()
+        return self._runnable_by_key[key]
+
+    def call_soon_dedup(self, key, fn):
+        """Enqueue a Python callable `fn` to be run soon. Use `key` to identify avoid duplication:
+        if `call_soon_dedup()` is called twice with the same `key`, we only ask Android for one
+        wake-up."""
+        if key in self._calling_soon:
+            return
+        self.handler.postDelayed(self.get_or_create_runnable(key, fn), 0)
+
+    def call_later(self, key, fn, timeout_millis):
+        """Enqueue a Python callable `fn` to be run after `timeout_millis` milliseconds. Since this
+        relies on a `java.lang.Runnable`, and we want create as few Java objects as possible, use
+        `key` as a cache key to avoid creating the same `Runnable` repeatedly."""
+        # Coerce timeout_millis to an integer since postDelayed() takes an integer (jlong).
+        self.handler.postDelayed(
+            self.get_or_create_runnable(key, fn), int(timeout_millis)
+        )
+
+
+# Some methods in this file are based on CPython's implementation.
```

Comment:
This should be at the top of the file.  We also don't need the *whole* license; just the text from clause 2 (and a link to the full license if we want to be nice).

Reply:
IMHO the license says we need the whole license, but I'll do what you say :)